### ISSUE TITLE
holes within some expressions are missed

### ISSUE THREAD

**Issue by coreyoconnor** [cb372/scala-typed-holes#10]
I have some code that contains `???` which is not recognized as a hole. In the sample below the holes in `parseCmdLine` do not result in output of the type while the one in `foo` does.

~~~
import caseapp._                                                                                                           
                                                                                                                           
object Main {                                                                                                              
  def main(args: Array[String]): Unit =                                                                                    
    println("Hello, world!")                                                                                               
                                                                                                                           
                                                                                                                           
  def parseCmdLine(args: List[String]): Option[Ops] = CaseApp.parse[Ops](args.toSeq) match {                               
    case Left(_)  => ???                                                                                                   
    case Right(_) => ???                                                                                                   
  }                                                                                                                        
                                                                                                                           
  def foo:Int = ???                                                                                                        
}                                                                                                                          
~~~

Not minimal yet. I'm filling this before I forget.

**Comment by cb372** [comment]
Thanks for the report! I’ll take a look. First things first I should really write some integration tests (#4) to capture stuff like this.

**Comment by coreyoconnor** [comment]
hmm. Off the top of my head I'm thinking the sbt scripted test harness could be used. That has a fairly nice interface for executing builds and checking the result. I have some experience with that and can see about making the above into a test if you agree.

* https://www.scala-sbt.org/1.x/docs/Testing-sbt-plugins.html

**Comment by cb372** [comment]
Maybe I'm misunderstanding the issue, but I wasn't able to reproduce this.

Compiling the following file:

```scala
object CaseApp {
  def parse[A](args: Seq[String]): Either[String, A] = Left("nope!")
}

case class Ops()

object Main {
  def main(args: Array[String]): Unit =
    println("Hello, world!")

  def parseCmdLine(args: List[String]): Option[Ops] = CaseApp.parse[Ops](args.toSeq) match {
    case Left(_)  => ???
    case Right(_) => ???
  }

  def foo:Int = ???
}
```

gives me the following compiler output, which is exactly what I would expect:

```scala
src/test/resources/issue-10/input.scala:12:
Found hole with type: Option[Ops]
Relevant bindings include
  args: List[String] (bound at input.scala:11:20)

    case Left(_)  => ???
                     ^
src/test/resources/issue-10/input.scala:13:
Found hole with type: Option[Ops]
Relevant bindings include
  args: List[String] (bound at input.scala:11:20)

    case Right(_) => ???
                     ^
src/test/resources/issue-10/input.scala:16: Found hole with type: Int
  def foo:Int = ???
                ^
```

Is this the kind of output you saw? If so, do you think there's some way in which it is incorrect or could be improved?

By the way, we now have tests 😄. The scenario for your issue is [here](https://github.com/cb372/scala-typed-holes/tree/investigate-issue-10/src/test/resources/issue-10). I thought about using sbt scripted but it's a bit slow because it has to boot sbt separately for each test case. In the end I hand-rolled a small test harness that compiles a .scala file with the plugin enabled and then compares the compiler output with an expected value.

**Comment by cb372** [comment]
I take it back! The output I pasted above is from the Scala 2.13.0-M5 compiler, but the holes in the pattern match don't show up when you compile with 2.11 or 2.12.

The tests have done their job perfectly - we now have a reproduction. I'll investigate further.

**Comment by coreyoconnor** [comment]
Cool! That makes sense. I neglected to mention I was using 2.11. Specifically, scala-native 0.38 on scala 2.11.

**Comment by cb372** [comment]
Released a fix for this in v0.0.5.
